### DISCUSSION TITLE
Tips for using perceptual precision?

### DISCUSSION THREAD

**Discussion by xinsight** [pointfreeco/swift-snapshot-testing#656] [Q&A]
I'm just trying out the 1.10.0 release with perceptual precision and it seems to have the same false positive issues as the existing precision parameter.

For example, these two images (which appear identical to me) pass with `.image(precision: 0.995)` but only pass with perceptual precision if I lower them to: `.image(perceptualPrecision: 0.962)`

![reference_1_1BFAA73A-05F7-45E2-94ED-7CAE06A00F11](https://user-images.githubusercontent.com/465971/194834633-4c2a9c94-db69-4745-bc55-4ebc23f327ff.png)
![failure_2_1BFAA73A-05F7-45E2-94ED-7CAE06A00F11](https://user-images.githubusercontent.com/465971/194834659-0592b84d-1a01-4d3c-b6db-9c3a2997f943.png)

The build-in diff shows as pure black, but in another diff tool, you can see there are differences in text and rounded corners.
<img width="397" alt="diff-that-shows-the-diff-red" src="https://user-images.githubusercontent.com/465971/194836746-95de43b9-d5b9-496d-bf4d-4ce968e10649.png">

Is perceptual precision just not great with dark colours? Any tips on how to use it? I was really hoping to crank everything up to 1.0 since the old precision diff often misses small issues with larger screenshots (since 1% of a large image is a lot of bad pixels!)

**Comment by kenmaz** [comment]
> .image(perceptualPrecision: 0.962)

How about specifying both `precision` and `perceptualPrecision`? Not specifying `precision` means specifying `precision = 1`.

https://github.com/Frameio/swift-snapshot-testing/blob/main/Sources/SnapshotTesting/Snapshotting/UIImage.swift#L68

In my understanding, it tests to see if areas with perceptual differences of less than `1-perceptualPrecision` % are less than `1-precision` % of the total pixels. Therefore, if `precision` is set to `1`, even a small perceptual difference is treated as a failure.
